File: package-rnaseq/bin/bdbag.pl
#! usr/bin/perl

=head1 NAME

  BDBAG CLOUD Version for AWS

=head1 VERSION

  Version 0.9
	Changes: *Added new functionality for AWS
		 *Added new flags for alignment
		 *add strict option to die if files dont exist
		 *Changed -all to no stop if folder doesnt exist
		 *Separated the ge and de commands
		 

=head1 DESCRIPTION
  Perl script to automate BDBag Generation for Report generation pipeline

=head1 SYNOPSIS
USAGE: 
  -p=path/to/outputrepository
  -o=path/to/store/bdbag

=head1 OPTIONS

B<--path, -p>
  Path to repository
B<--out, -o>
  Path to output location
B<--pid, -pid>
  Pipeline ID
B<--name, -n>
  Name of project
B<--align, -align>
  Pull Alignment files only
B<--fastqc, -fqc>
  Pull only FastQC files
B<--ge, -ge>
  Pull GE files only
B<--de, -de>
  Pull DE files only
B<--cloud, -cloud>
  Pull xml files
B<--include, -include-files>
  sam or bam or sam,bam 
B<--includesam, -sam>
  Include SAM files
B<--includebam, -bam>
  Include BAM files
B<--all, -all>
  Pull all files from this pipeline
B<--rb, -rb>
  Use this r binary
B<--rs, -rs>
  Use this bdbag.R script 
B<-frs, --filers>
  Use this rename_de.R script
B<-crs, --countsrs>
  Use this Generate_all_counts.R script
B<--rp,-rp>
  Use these r parameters
B<--makebag, -make>
  Make BD bag
B<--update, -update>
  Update Bag


=head1 CONTACT  

  Apaala Chatterjee at achatterjee@som.umaryland.edu
  
=head1 EXAMPLE

perl bdbag_generator.pl -p /local/projects-t3/XYZ/rhl7/rnaseq/ergatis -o /local/projects/RNASEQ/Report_Generation/Beta/ -pid 13898000001 -n XYZ -fqc 

perl bdbag_generator.pl -p /local/projects-t3/XYZ/rhl7/rnaseq/ergatis -o /local/projects/RNASEQ/Report_Generation/Beta/ -pid 13898000002 -n XYZ -ge -de

perl bdbag_generator.pl -p /local/projects-t3/XYZ/rhl7/rnaseq/ergatis -o /local/projects/RNASEQ/Report_Generation/Beta/ -n XYZ -cloud -bam -sam -make
=cut

use strict;
use warnings;
use Getopt::Long qw(:config no_ignore_case no_auto_abbrev pass_through);
use English qw( -no_match_vars );
use File::Basename;
use Pod::Usage;
use Cwd;
use FindBin qw($RealBin);
use Cwd 'abs_path';
use MIME::Lite;

####################################################################
###                             User Input
#####################################################################

my $results = GetOptions ('path|p=s' => \my $path,
                          'out|o=s' => \my $outpath,
			  'pid|pid=s' => \my $pid,
			  'name|n=s' => \my $bag,
			  'fastqc|fqc' => \my $dofqc,
			  'ge|ge' => \my $doge,
			  'align|align' => \my $doalign,
			  'e|email' => \my $email,
			  'de|de' => \my $dode,
			  'all|all' => \my $all,
			  'cloud|cloud'=> \my $cloud,
			  'include|include-files=s' => \my $include,
			  'bam|includebam' => \my $bam,
			  'sam|includesam' => \my $sam,
			  'rb|rb=s' => \my $rb,
			  'rp|rp=s' => \my $rp,
			  'rs|rs=s' => \my $rs,
			  'crs|countsrs=s' => \my $crs,
			  'frs|filers=s' => \my $frs,
			  'makebag|make' => \my $makebag,
			  'strict|strict' => \my $strict,
			  'update|u' => \my $update,
			  'help|h' => \my $help)|| pod2usage();

if( $help ){
  pod2usage( {-exitval => 0, -verbose => 2, -output => \*STDERR} );
}

###############################################################
## MAIN
################################################################
chdir $outpath;
my $syscmd="";
print("Setting directory to: ");
system 'pwd';
#print(defined $rb);

if (! defined $rb )
{
$rb="/usr/local/packages/r-3.4.0/bin/Rscript";
#print($rb);
}

if (! defined $rs )
{

$rs=$RealBin."/bdbag.R";
#print($rp);
}

if (! defined $crs )
{

$crs=$RealBin."/Generate_all_counts.R";
#print($rp);
}

if (! defined $frs )
{
$frs=$RealBin."/rename_de.R";
#print($rp);
}

if (! defined $rp )
{
$rp="";
#print($rs);
}

#my $bag="bdBag".$pid;
if (! -e $bag) {
 mkdir($bag) or die "Can't create /path/to/dir:$!\n";
}

###############################################################
### Directory set up
################################################################
chdir $bag;
my $rep_path=$path."/output_repository";

if($doalign)
{
my $align=$rep_path."/wrapper_align/".$pid."_wrap/Summary.txt";                                                                                                                                                                       
my $syscmd="cp ".$align." .";                                                                                                                                                                                                         
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";    
}


###############################################################
#### bdbag report
###############################################################
my $timestamp = getLoggingTime();
my $filename1 = 'bdbag_details.txt';
open(my $fh, '>>', $filename1) or die "Could not open file '$filename1' $!";
my $str1=$rb."\n".$rs."\n".$crs."\n".$frs."\n".$rep_path."\n".$bag."\n".$timestamp."\n";
print $fh $str1;
close $fh;

###############################################################
#### GE Report
###############################################################

if($doge)
{
	print "Copying and Generating GE Files \n";
	my $checkrpkm=$rep_path."/rpkm_coverage_stats/".$pid."_rpkm_cvg";
	if(-e $checkrpkm)
	{

        	print("\nCopying RPMK Files \n");
        	if (! -e 'RPKM') {
                	mkdir('RPKM') or die "Can't create /path/to/dir:$!\n";
        	}

        	my $ge=$rep_path."/rpkm_coverage_stats/".$pid."_rpkm_cvg/i1/g*/genic_coverage/*.txt";
        	$syscmd="cp ".$ge." RPKM";
        	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	}

	if (! -e 'Counts') 
	{                                                                                                                                       
 		mkdir('Counts') or die "Can't create /path/to/dir:$!\n";                                                                                                  
	}                                                                                                
	my $counts=$rep_path."/htseq/".$pid."*_counts/i1/g*/*.counts";                                                                                             
	$syscmd="cp ".$counts." Counts";                                                                                                                           
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n"; 
	$syscmd=$rb." ".$crs;
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	#print($syscmd);
	#system("/usr/local/packages/r-3.4.0/bin/Rscript /home/apaala.chatterjee/RNA_Report/Generate_all_counts.R");
}

###############################################################
##### DE Report
################################################################

if($dode)
{
	if (! -e 'DE') 
	{
 		mkdir('DE') or die "Can't create /path/to/dir:$!\n";
	}
	if(! -e 'OutputTables/logCPM_RPKM.txt')
	{
		my $copyCounts=$rep_path."/deseq/".$pid."_differential_expression/i1/g*/";
		#print($copyCounts);
		my $copysyscmd=$rb." ".$frs." ".$copyCounts." DE";
		#print($copysyscmd);
		system($copysyscmd) == 0 or die "system($syscmd) failed:$?\n";
	}

	my $de=$rep_path."/deseq/".$pid."_differential_expression/i1/g*/*de_genes.txt";
	$syscmd="cp ".$de." DE";
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
}

###############################################################
###### FASTQC Report
###############################################################

if($dofqc)
{
	if (! -e 'Files')
	{
 		mkdir('Files') or die "Can't create /path/to/dir:$!\n";
	}
	if (! -e 'Files/KmerProfiles') 
	{
 		mkdir('Files/KmerProfiles') or die "Can't create /path/to/dir:$!\n";
	}
	if (! -e 'Files/BaseQuality') 
	{
 		mkdir('Files/BaseQuality') or die "Can't create /path/to/dir:$!\n";
	}
	print "Copying FASTQC Images\n";
	my $fqc= $rep_path."/fastqc_stats/".$pid."_fastqc/i1/g*/*/Images/";
	#print $fqc;
	$syscmd="cp ".$fqc."*_sequence.kmer_profiles.png Files/KmerProfiles/";
	#print $syscmd;
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	$syscmd="cp ".$fqc."*_base_quality.png Files/BaseQuality/"; 
	#print $syscmd;                                                                                                                                                                                                                                                                               
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
}


###############################################################
####### Strict Options
###############################################################


if($strict)
{
print "Generating Directory structure \n";
if (! -e 'Files') {
 mkdir('Files') or die "Can't create /path/to/dir:$!\n";
}
if (! -e 'Files/KmerProfiles') {
 mkdir('Files/KmerProfiles') or die "Can't create /path/to/dir:$!\n";
}
if (! -e 'Files/BaseQuality') {                                                                                                                                                                                                                                                               
 mkdir('Files/BaseQuality') or die "Can't create /path/to/dir:$!\n";                                                                                                                                                                                                                          
}
print "Copying FASTQC Images\n";                                                                                                                                                                                                                                                              
my $fqc= $rep_path."/fastqc_stats/".$pid."_fastqc/i1/g*/*/Images/";                                                                                                                                                                                                                               
$syscmd="cp ".$fqc."*_sequence.kmer_profiles.png Files/KmerProfiles/";                                                                                                                                                                                                                                 
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";                                                                                                                                                                                                                                    
$syscmd="cp ".$fqc."*_base_quality.png Files/BaseQuality/";                                                                                                                                                                                                                                   
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";

my $align=$rep_path."/wrapper_align/".$pid."_wrap/Summary.txt";
$syscmd="cp ".$align." .";
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";

if (! -e 'DE') {
 mkdir('DE') or die "Can't create /path/to/dir:$!\n";
}

if(! -e 'OutputTables/logCPM_RPKM.txt')
        {
                my $copyCounts=$rep_path."/deseq/".$pid."_differential_expression/i1/g*/";
                my $copysyscmd=$rb." ".$frs." ".$copyCounts." DE";
                system($copysyscmd) == 0 or die "system($syscmd) failed:$?\n";
        }
if (! -e 'RPKM') {
 mkdir('RPKM') or die "Can't create /path/to/dir:$!\n";
}
print "Generating GE Files \n";
my $ge=$rep_path."/rpkm_coverage_stats/".$pid."_rpkm_cvg/i1/g*/genic_coverage/*.txt";
$syscmd="cp ".$ge." RPKM";
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
my $de=$rep_path."/deseq/".$pid."_differential_expression/i1/g*/*de_genes.txt";
$syscmd="cp ".$de." DE";
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
if (! -e 'Counts') {
 mkdir('Counts') or die "Can't create /path/to/dir:$!\n";
}
my $counts=$rep_path."/htseq/".$pid."*_counts/i1/g*/*.counts";
$syscmd="cp ".$counts." Counts";
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
$syscmd=$rb." ".$crs;
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
#system("/usr/local/packages/r-3.4.0/bin/Rscript /home/apaala.chatterjee/RNA_Report/Generate_all_counts.R");
}

###############################################################
######## All Option
################################################################


if($all)
{
print "Generating Directory structure \n";
if (! -e 'Files') {
 mkdir('Files') or die "Can't create /path/to/dir:$!\n";
}
if (! -e 'Files/KmerProfiles') {
 mkdir('Files/KmerProfiles') or die "Can't create /path/to/dir:$!\n";
}
if (! -e 'Files/BaseQuality') {                                                                                                                                                                                                                           
 mkdir('Files/BaseQuality') or die "Can't create /path/to/dir:$!\n";                                                                                                                                                                                      
}

####FastQC check
print "Copying FASTQC Images\n";                                                                                                                                                                                                                          
my $fqc= $rep_path."/fastqc_stats/".$pid."_fastqc/i1/g*/*/Images/";                                                                                                                                                                                      
my $fqc_check= $rep_path."/fastqc_stats/".$pid."_fastqc/";
if( -d $fqc_check)
{
#base sequence content instead of Kmer
#change the fastqc option to 11.7v of fastqc in config file
#fastqc.pl script change


$syscmd="cp ".$fqc."*_sequence.kmer_profiles.png Files/KmerProfiles/";                                                                                                                                                                                             
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";                                                                                                                                                                                                
$syscmd="cp ".$fqc."*_base_quality.png Files/BaseQuality/";                                                                                                                                                                                               
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
}
else
{
print("FastQC Files not Found at $fqc \n Proceeding to next step \n");
print(-d $fqc);
}

###Alignment Check
my $align=$rep_path."/wrapper_align/".$pid."_wrap/Summary.txt";
if(-e $align)
{
$syscmd="cp ".$align." .";
print("Copying Summary.txt from $align \n");
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
}
else
{
print("Summary.txt not found at $align \n Proceeding to next step \n");
}
if (! -e 'DE') {
 mkdir('DE') or die "Can't create /path/to/dir:$!\n";
}

#print "Generating GE DE Files \n";
#####GE
my $checkrpkm=$rep_path."/rpkm_coverage_stats/".$pid."_rpkm_cvg";
my $ge=$rep_path."/rpkm_coverage_stats/".$pid."_rpkm_cvg/i1/g*/genic_coverage/*.txt";
my $ge_check=$rep_path."/rpkm_coverage_stats/".$pid."_rpkm_cvg/i1/";
if(-d $ge_check)
{
print "Generating GE Files \n";
if(-e $checkrpkm)
{

        print("\nCopying RPMK Files \n");
        if (! -e 'RPKM') {
                mkdir('RPKM') or die "Can't create /path/to/dir:$!\n";
        }

        my $ge=$rep_path."/rpkm_coverage_stats/".$pid."_rpkm_cvg/i1/g*/genic_coverage/*.txt";
        $syscmd="cp ".$ge." RPKM";
        system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
}

#$syscmd="cp ".$ge." RPKM";
#system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
}
else
{
print("GE Files not found at $ge_check \n Proceeding with next step \n");
}
my $de=$rep_path."/deseq/".$pid."_differential_expression/i1/g*/*de_genes.txt";
my $de_check=$rep_path."/deseq/".$pid."_differential_expression/i1/";
if(-d $de_check)
{
print "Generating DE Files \n";
$syscmd="cp ".$de." DE";
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
}
else
{
print("DE Files not Found at $de_check \n Proceeding to next step \n");
}

if (! -e 'Counts') {
 mkdir('Counts') or die "Can't create /path/to/dir:$!\n";
}
###Check Counts
my $count_check=$rep_path."/htseq/".$pid."_exon_counts/i1/";
my $counts=$rep_path."/htseq/".$pid."*_counts/i1/g*/*.counts";
if (-e $count_check)
{
print("Copying Count File and generating all counts\n");
$syscmd="cp ".$counts." Counts";
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
$syscmd=$rb." ".$crs;
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
#system("/usr/local/packages/r-3.4.0/bin/Rscript /home/apaala.chatterjee/RNA_Report/Generate_all_counts.R");
}
else
{
print("Counts files not found at $count_check\n");
}
}

###############################################################
######## CLOUD Option-xml files
################################################################

if ($cloud)
{

	my $ppath=$outpath."/".$bag;
	print("Setting directory to $ppath \n");
	chdir($ppath);
		if (! -e 'XML')
		{
			 mkdir('XML') or die "Can't create /path/to/dir:$!\n";
		}
	print("Gathering files for cloud and putting then in $outpath \n");
	my $pipelineXML=$path."workflow/runtime/pipeline/".$pid."/pipeline.xml";
	if(-e $pipelineXML)
	{
		print("Copying pipeline.xml from $pipelineXML\n");
		$syscmd="cp ".$pipelineXML." XML";
		system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	}
	$syscmd="ls ".$path."/workflow/runtime/*/".$pid."*/component.xml > comp.list";
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	$syscmd="ls ".$path."/workflow/runtime/*/".$pid."*/i1/*.xml.gz > i1.list";
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	$syscmd="ls ".$path."/workflow/runtime/*/".$pid."*/i1/g*/g*.xml.gz > g_all.list";
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	my $componentXML=$path."workflow/runtime/wrapper_align/".$pid."_wrap/component.xml";
	$syscmd=$rb."/Rscript ".$rp." ".$rs." ".$path." ".$ppath."/ ".$pid." cloud";
	print("\n");
	print($rs);
	print("\n");
	print($syscmd);
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
}

if(defined($include))
{
if($include eq "bam")
{
	print("Copying bam files");
	my $ppath=$outpath."/".$bag;
	chdir($ppath);
	print("current directory is $ppath");
	if (! -e 'XML')
        {
                 mkdir('XML') or die "Can't create /path/to/dir:$!\n";
        }
	$syscmd="ls ".$path."/output_repository/samtools_file_convert/".$pid."_sorted_name/i1/g*/*sorted_by_name.bam > sorted_by_name.bam.list";
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	$syscmd="ls ".$path."/output_repository/samtools_file_convert/".$pid."_sorted_position/i1/g*/*sorted_by_position.bam  > sorted_by_position.bam.list";
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
#$syscmd="/usr/local/packages/r-3.3.1/bin/Rscript /home/apaala.chatterjee/AWS/bdbag.R ".$path." ".$ppath."/ ".$pid." bam";

	$syscmd=$rb."/Rscript ".$rp." ".$rs." ".$path." ".$ppath."/ ".$pid." bam";
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	print($syscmd);
}

if($include eq "sam")
{
	print("Copying sam files");
	my $ppath=$outpath."/".$bag;
	chdir($ppath);
	if (! -e 'XML')
        {
                 mkdir('XML') or die "Can't create /path/to/dir:$!\n";
        }

	$syscmd="ls ".$path."/output_repository/samtools_file_convert/".$pid."_sorted_name/i1/g*/*sorted_by_name.sam > sorted_by_name.sam.list";
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	#print($syscmd);

#$syscmd="/usr/local/packages/r-3.3.1/bin/Rscript /home/apaala.chatterjee/AWS/bdbag.R ".$path." ".$ppath."/ ".$pid." sam";
	$syscmd=$rb."/Rscript ".$rp." ".$rs." ".$path." ".$ppath."/ ".$pid." sam";
	system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
}

if($include eq "bam,sam")
{
	print("Copying SAM and BAM files");
	my $ppath=$outpath."/".$bag;
        chdir($ppath);
        if (! -e 'XML')
        {
                 mkdir('XML') or die "Can't create /path/to/dir:$!\n";
        }
        $syscmd="ls ".$path."/output_repository/samtools_file_convert/".$pid."_sorted_name/i1/g*/*sorted_by_name.bam > sorted_by_name.bam.list";
        system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
        $syscmd="ls ".$path."/output_repository/samtools_file_convert/".$pid."_sorted_position/i1/g*/*sorted_by_position.bam  > sorted_by_position.bam.list";
        system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	$syscmd=$rb."/Rscript ".$rp." ".$rs." ".$path." ".$ppath."/ ".$pid." bam";
        system($syscmd) == 0 or die "system($syscmd) failed:$?\n";


	$syscmd="ls ".$path."/output_repository/samtools_file_convert/".$pid."_sorted_name/i1/g*/*sorted_by_name.sam > sorted_by_name.sam.list";
        system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
	$syscmd=$rb."/Rscript ".$rp." ".$rs." ".$path." ".$ppath."/ ".$pid." sam";
        system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
}
}

if($makebag)
{
print("Setting directory to $outpath \n");
chdir($outpath);
print("Making Bag: \n");
$syscmd="bdbag ".$bag." --archiver zip";
print ($syscmd);
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";   
}

if($update)
{
print "Updating Bag \n";
chdir($outpath);
$syscmd="bdbag ".$bag." --archiver zip --update";
system($syscmd) == 0 or die "system($syscmd) failed:$?\n";
}

sub getLoggingTime {

    my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst)=localtime(time);
    my $nice_timestamp = sprintf ( "%04d%02d%02d %02d:%02d:%02d",
                                   $year+1900,$mon+1,$mday,$hour,$min,$sec);
    return $nice_timestamp;
}

File: package-rnaseq/bin/bdbag.R
#!/usr/local/packages/r-3.4.0/bin Rscript
library(rmarkdown)
library(knitr)
args = commandArgs(trailingOnly=TRUE)
ergatis=args[1]
output=args[2]
#listf=args[3]
pid=args[3]
#stopifnot(length(args) ==4)
#stopifnot(length(args) ==4)
#print(output)
#listf<-read.table(paste0(dir,"comp.list"))
#comp<-read.table(listf)
workpath<-paste0(ergatis,"/workflow/runtime/")
workpath<-gsub("//","/",workpath)
#workpath
#comp<-comp$V1
#head(comp)
replacepid<-paste0(pid,".*")
#replacepid
mainDir<-paste0(output,"/XML/")
#mainDir
#print("Starting Dir creation")
#args


#######################
#####COPY COMPONENT XML
#####################listf<-read.table(paste0(dir,"comp.list"))
if("cloud" %in% (args)){
comp<-read.table(paste0(output,"comp.list"))
comp<-comp$V1
for( i in comp)
{
  #print(i)
  str1<-i
  str1<-gsub("//","/",str1)
  #print(str1)
  str2<-gsub(workpath,"", str1)
  #print(str2)
  str3<-gsub(replacepid,"",str2)
  #print(str3)
  #mainDir<-output
  dir.create(file.path(mainDir, str3), showWarnings = FALSE)
  Nmaindir<-paste0(mainDir,str3)
  #print("NMaindir")
  #print(Nmaindir)
  dir.create(file.path(Nmaindir,"i1"),showWarnings = FALSE)
  copypath<-file.copy(str1, paste0(Nmaindir,"/","i1/"))
  #print( paste0(Nmaindir,"/","i1/"))
}


#######################
#####COPY i1.xml.gz
#######################
print ("Copying XML files as directed")

dir<-gsub("/XML","", output)
i1list<-read.table(paste0(dir,"i1.list"))
#i1list
output
i1list<-i1list$V1
for(j in i1list)
{
	str1<-j
	str1<-gsub("//","/",str1)
	str2<-gsub(workpath,"", str1)
	str3<-gsub(replacepid,"",str2)
#	print(str3)
	cpdir<-paste0(output,str3,"/i1")
#	print(cpdir)
	#dir.create(file.path(Nmaindir,"i1"),showWarnings = FALSE)
	copypath<-file.copy(str1, cpdir)
}

gall<-read.table(paste0(dir,"g_all.list"))
gall<-gall$V1
replaceg<-paste0(pid,".*/i")
for(k in gall)
{
	str1<-k
	str1<-gsub("//","/",str1)
	if(grepl("mate",str1)==TRUE)
	{
	str2<-gsub(workpath,"", str1)
	#print(str2)
	str3<-gsub(paste0(pid,"_"),"/i1/",str2)
	#print(str3)
	str4<-gsub("i1/g","g",str3);
	#print(str4)
	xmlname<-basename(str4)
	cpto<-gsub(xmlname, "", str4)
	#print(cpto)
	str5<-strsplit(cpto, "/(?=[^,]+$)", perl=TRUE)
        str5<-unlist(str5[[1]])
        len<-length(str5)
        gnum<-str5[len]
        #print(gnum)	
	matedir<-gsub(gnum,"",cpto)
	#print(matedir)
	splitM<-strsplit(matedir, "/(?=[^,]+$)", perl=TRUE)
	splitM<-unlist(splitM[[1]])
	lenM<-length(splitM)
	mateName<-splitM[lenM]
	#print(mateName)
	mateDir<-gsub(mateName,"",matedir)
	#print(mateDir)
	mkMate<-paste0(output,mateDir)
	#print(mkMate)	
	matedir1<-paste0(output,matedir)
	matedir1<-gsub("//","/",matedir1)
	#print(matedir1)
	dir.create(file.path(mkMate,mateName),showWarnings = FALSE)	
	dir.create(file.path(matedir1,gnum),showWarnings = FALSE)
	cpg<-paste0(matedir1,gnum)
	#print(cpg)
	copypath<-file.copy(str1,cpg)
	}
	if(grepl("mate",str1)==FALSE)
	{	
	str2<-gsub(workpath,"", str1)
	str3<-gsub(replaceg,"i",str2)
	#print(str3)
	xmlname<-basename(str3)
	cpto<-gsub(xmlname, "", str3)
	fullto<-paste0(mainDir,cpto)
	fulltoC<-gsub("//","/",fullto)
	fullto<-gsub("/g.*", "",fulltoC)
	#print(fullto)
	str4<-strsplit(cpto, "/(?=[^,]+$)", perl=TRUE)
	str4<-unlist(str4[[1]])
	len<-length(str4)
	gnum<-str4[len]
	#print(gnum)
	dir.create(file.path(fullto, gnum), showWarnings = FALSE)
	copypath<-file.copy(str1, fulltoC)
	#print(cpto)
	}
}
}

####SAM and BAM Files

if("bam" %in% (args))
{

print("Bam files being copied")

fpath<-gsub("/XML","",output)
print("This is what fpath variable has")
print(fpath)
dir.create(file.path(fpath, "Files"), showWarnings = FALSE)
fworking<-paste0(fpath,"Files/")
print(fworking)
bam_path_name<-paste0(fpath,"/sorted_by_name.bam.list")
bam_path_name<-gsub("//","/",bam_path_name)
print(bam_path_name)
bam<-read.table(bam_path_name)
bam<-bam$V1
#testprint<-paste0(fpath,"/sorted_by_position.bam.list")
#print(bam)
bam_position_name<-paste0(fpath,"/sorted_by_position.bam.list")
bam_position_name<-gsub("//","/",bam_position_name)
positionbam<-read.table(bam_position_name)
positionbam<-positionbam$V1

dir.create(file.path(fworking, "sorted_by_name"), showWarnings = FALSE)
dir.create(file.path(fworking, "sorted_by_position"), showWarnings = FALSE)
bamdir<- paste0(fworking,"sorted_by_name")
posdir<-paste0(fworking,"sorted_by_position")

bworkpath<-paste0(ergatis,"output_repository")
bworkpath<-gsub("//","/",bworkpath)
#bworkpath

for(l in bam)
{
print("Starting with sorted by name")
str1<-l
str1<-gsub("//","/",str1)
str2<-gsub(paste0(pid,".*/g"), "g", str1)
str2<-gsub(bworkpath,"",str2)
str3<-gsub("/samtools_file_convert/","/",str2)
fname<-basename(str3)
gnum<-gsub(fname,"",str3)
gnum<-basename(gnum)
dir.create(file.path(bamdir, gnum), showWarnings = FALSE)
bampath<-paste0(bamdir,"/", gnum)
copypath<-file.copy(l, bampath)

}

for(m in positionbam)
{
str1<-m
str1<-gsub("//","/",str1)
str2<-gsub(paste0(pid,".*/g"), "g", str1)
str2<-gsub(bworkpath,"",str2)
str3<-gsub("/samtools_file_convert/","/",str2)
fname<-basename(str3)
gnum<-gsub(fname,"",str3)
gnum<-basename(gnum)
dir.create(file.path(posdir, gnum), showWarnings = FALSE)
pospath<-paste0(posdir,"/", gnum)
#print(pospath)
copypath<-file.copy(m, pospath)

}

}

if("sam" %in% (args))
{
print("SAM files being copied");
fpath<-gsub("/XML","",output)
fpath
dir.create(file.path(fpath, "Files"), showWarnings = FALSE)
fworking<-paste0(fpath,"Files/")
fworking
sam<-read.table("sorted_by_name.sam.list")
sam<-sam$V1
head(sam)
dir.create(file.path(fworking, "sorted_by_name"), showWarnings = FALSE)
samdir<- paste0(fworking,"sorted_by_name")
bworkpath<-paste0(ergatis,"output_repository")

for(n in sam)
{
str1<-n
str1<-gsub("//","/",str1)
#print(str1)
str2<-gsub(paste0(pid,".*/g"), "g", str1)
str2<-gsub(bworkpath,"",str2)
str3<-gsub("/samtools_file_convert/","/",str2)
fname<-basename(str3)
gnum<-gsub(fname,"",str3)
gnum<-basename(gnum)
dir.create(file.path(samdir, gnum), showWarnings = FALSE)
sampath<-paste0(samdir,"/", gnum)
#print(sampath)
copypath<-file.copy(n, sampath)
#print(gnum)
#print(str3)

}


}
